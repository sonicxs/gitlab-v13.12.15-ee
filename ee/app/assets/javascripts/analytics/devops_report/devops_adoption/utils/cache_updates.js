import produce from 'immer';
import devopsAdoptionSegmentsQuery from '../graphql/queries/devops_adoption_segments.query.graphql';

export const addSegmentsToCache = (store, segments, variables) => {
  const sourceData = store.readQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
  });

  const data = produce(sourceData, (draftData) => {
    draftData.devopsAdoptionSegments.nodes = [
      ...draftData.devopsAdoptionSegments.nodes,
      ...segments,
    ];
  });

  store.writeQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
    data,
  });
};

export const deleteSegmentsFromCache = (store, segmentIds, variables) => {
  const sourceData = store.readQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
  });

  const updatedData = produce(sourceData, (draftData) => {
    draftData.devopsAdoptionSegments.nodes = draftData.devopsAdoptionSegments.nodes.filter(
      ({ id }) => !segmentIds.includes(id),
    );
  });

  store.writeQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
    data: updatedData,
  });
};
