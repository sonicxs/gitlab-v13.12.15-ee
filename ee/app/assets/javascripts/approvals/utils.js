import { flatten } from 'lodash';

export const joinRuleResponses = (responsesArray) =>
  Object.assign({}, ...responsesArray, {
    rules: flatten(responsesArray.map(({ rules }) => rules)),
  });
