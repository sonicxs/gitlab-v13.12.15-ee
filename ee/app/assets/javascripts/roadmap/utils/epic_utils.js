import createGqClient, { fetchPolicies } from '~/lib/graphql';

export const gqClient = createGqClient(
  {},
  {
    fetchPolicy: fetchPolicies.NO_CACHE,
  },
);

export const addIsChildEpicTrueProperty = (obj) => ({ ...obj, isChildEpic: true });

export const generateKey = (epic) => `${epic.isChildEpic ? 'child-epic-' : 'epic-'}${epic.id}`;
