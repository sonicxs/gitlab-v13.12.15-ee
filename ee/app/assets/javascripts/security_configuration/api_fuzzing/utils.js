/* eslint-disable @gitlab/require-i18n-strings */
import { isString } from 'lodash';

export const insertTip = ({ snippet, tip, token }) => {
  if (!isString(snippet)) {
    throw new Error('snippet must be a string');
  }
  if (!isString(tip)) {
    throw new Error('tip must be a string');
  }
  if (!isString(token)) {
    throw new Error('token must be a string');
  }
  const lines = snippet.split('\n');
  for (let i = 0; i < lines.length; i += 1) {
    if (lines[i].includes(token)) {
      const indent = lines[i].match(/^[ \t]+/)?.[0] ?? '';
      lines[i] = lines[i].replace(token, `# ${tip}\n${indent}${token}`);
      break;
    }
  }
  return lines.join('\n');
};

export const insertTips = (snippet, tips = []) =>
  tips.reduce(
    (snippetWithTips, { tip, token }) => insertTip({ snippet: snippetWithTips, tip, token }),
    snippet,
  );
