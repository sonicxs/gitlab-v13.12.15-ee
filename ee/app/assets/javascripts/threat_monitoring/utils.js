/**
 * Get the height of the wrapper page element
 * This height can be used to determine where the highest element goes in a page
 * Useful for gl-drawer's header-height prop
 * @param {String} class the content wrapper class
 * @returns {String} height in px
 */
export const getContentWrapperHeight = (contentWrapperClass) => {
  const wrapperEl = document.querySelector(contentWrapperClass);
  return wrapperEl ? `${wrapperEl.offsetTop}px` : '';
};
