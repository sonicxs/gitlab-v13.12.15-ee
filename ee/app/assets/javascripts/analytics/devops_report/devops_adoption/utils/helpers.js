import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the segments,
 *
 * @param {Object} params the segment data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ segments, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!segments.length) {
    return true;
  }

  const anyPendingSegments = segments.some((node) => node.latestSnapshot === null);
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingSegments || dataNotRefreshedToday;
};
