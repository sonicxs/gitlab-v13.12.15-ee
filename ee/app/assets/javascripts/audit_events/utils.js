import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { AVAILABLE_TOKEN_TYPES, AUDIT_FILTER_CONFIGS } from './constants';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));

  return {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_id: entityValue?.value.data,
    entity_type: getEntityTypeFromType(entityValue?.type),
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };
};
