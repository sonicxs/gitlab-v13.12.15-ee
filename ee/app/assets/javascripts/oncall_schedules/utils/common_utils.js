import { newDateAsLocaleTime } from '~/lib/utils/datetime_utility';
import { sprintf, __ } from '~/locale';
import { ASSIGNEE_COLORS_COMBO } from '../constants';

/**
 * Returns formatted timezone string, e.g. (UTC-09:00) AKST Alaska
 *
 * @param {Object} tz
 * @param {String} tz.name
 * @param {String} tz.formatted_offset
 * @param {String} tz.abbr
 *
 * @returns {String}
 */
export const getFormattedTimezone = (tz) => {
  return sprintf(__('(UTC %{offset}) %{timezone}'), {
    offset: tz.formatted_offset,
    timezone: `${tz.abbr} ${tz.name}`,
  });
};

/**
 * Returns `true` for non-empty string, otherwise returns `false`
 *
 * @param {String} startDate
 *
 * @returns {Boolean}
 */
export const isNameFieldValid = (nameField) => {
  return Boolean(nameField?.length);
};

/**
 * Returns a Array of Objects that represent the shift participant
 * with his/her username and unique shift color values
 *
 * @param {Object[]} participants
 * @param {string} participants[].username - The username of the participant.
 *
 * @returns {Object[]} A list of values to save each participant
 * @property {string} username
 * @property {string} colorWeight
 * @property {string} colorPalette
 */
export const getParticipantsForSave = (participants) =>
  participants.map(({ username }, index) => {
    const colorIndex = index % ASSIGNEE_COLORS_COMBO.length;
    const { colorWeight, colorPalette } = ASSIGNEE_COLORS_COMBO[colorIndex];

    return {
      username,
      colorWeight,
      colorPalette,
    };
  });

/**
 * Parses a activePeriod string into an integer value
 *
 * @param {String} hourString
 */
export const parseHour = (hourString) => parseInt(hourString.slice(0, 2), 10);

/**
 * Parses a rotation date for use in the add/edit rotation form
 *
 * @param {ISOString} dateTimeString
 * @param {Timezone string - long} scheduleTimezone
 */
export const parseRotationDate = (dateTimeString, scheduleTimezone) => {
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    hourCycle: 'h23',
    timeZone: scheduleTimezone,
    timeZoneName: 'long',
  };
  const formatter = new Intl.DateTimeFormat('en-US', options);
  const parts = formatter.formatToParts(Date.parse(dateTimeString));
  const [month, , day, , year, , hour] = parts.map((part) => part.value);
  // The datepicker uses local time
  const date = newDateAsLocaleTime(`${year}-${month}-${day}`);
  const time = parseInt(hour, 10);

  return { date, time };
};
