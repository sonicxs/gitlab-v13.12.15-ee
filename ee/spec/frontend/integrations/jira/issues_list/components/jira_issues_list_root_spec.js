import { shallowMount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';

import JiraIssuesListRoot from 'ee/integrations/jira/issues_list/components/jira_issues_list_root.vue';
import waitForPromises from 'helpers/wait_for_promises';
import createFlash from '~/flash';
import IssuableList from '~/issuable_list/components/issuable_list_root.vue';
import axios from '~/lib/utils/axios_utils';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import httpStatus from '~/lib/utils/http_status';

import { mockProvide, mockJiraIssues } from '../mock_data';

jest.mock('~/flash');
jest.mock('~/issuable_list/constants', () => ({
  DEFAULT_PAGE_SIZE: 2,
  IssuableStates: jest.requireActual('~/issuable_list/constants').IssuableStates,
  IssuableListTabs: jest.requireActual('~/issuable_list/constants').IssuableListTabs,
  AvailableSortOptions: jest.requireActual('~/issuable_list/constants').AvailableSortOptions,
}));

const resolvedValue = {
  headers: {
    'x-page': 1,
    'x-total': 3,
  },
  data: mockJiraIssues,
};

describe('JiraIssuesListRoot', () => {
  let wrapper;
  let mock;

  const findIssuableList = () => wrapper.find(IssuableList);

  const createComponent = ({ provide = mockProvide, initialFilterParams = {} } = {}) => {
    wrapper = shallowMount(JiraIssuesListRoot, {
      propsData: {
        initialFilterParams,
      },
      provide,
    });
  };

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    wrapper.destroy();
    mock.restore();
  });

  describe('on mount', () => {
    describe('while loading', () => {
      it('sets issuesListLoading to `true`', async () => {
        jest.spyOn(axios, 'get').mockResolvedValue(new Promise(() => {}));

        createComponent();

        await wrapper.vm.$nextTick();

        const issuableList = findIssuableList();
        expect(issuableList.props('issuablesLoading')).toBe(true);
      });

      it('calls `axios.get` with `issuesFetchPath` and query params', () => {
        jest.spyOn(axios, 'get');
        createComponent();

        expect(axios.get).toHaveBeenCalledWith(
          mockProvide.issuesFetchPath,
          expect.objectContaining({
            params: {
              with_labels_details: true,
              page: wrapper.vm.currentPage,
              per_page: wrapper.vm.$options.defaultPageSize,
              state: wrapper.vm.currentState,
              sort: wrapper.vm.sortedBy,
              search: wrapper.vm.filterParams.search,
            },
          }),
        );
      });
    });

    describe('when request succeeds', () => {
      beforeEach(async () => {
        jest.spyOn(axios, 'get').mockResolvedValue(resolvedValue);

        createComponent();
        await waitForPromises();
      });

      it('sets `currentPage` and `totalIssues` from response headers and `issues` & `issuesCount` from response body when request is successful', async () => {
        const issuableList = findIssuableList();
        const issuablesProp = issuableList.props('issuables');

        expect(issuableList.props()).toMatchObject({
          currentPage: resolvedValue.headers['x-page'],
          previousPage: resolvedValue.headers['x-page'] - 1,
          nextPage: resolvedValue.headers['x-page'] + 1,
          totalItems: resolvedValue.headers['x-total'],
        });
        expect(issuablesProp).toHaveLength(mockJiraIssues.length);

        const firstIssue = convertObjectPropsToCamelCase(mockJiraIssues[0], { deep: true });
        expect(issuablesProp[0]).toEqual({
          ...firstIssue,
          id: 31596,
          author: {
            ...firstIssue.author,
            id: 0,
          },
        });
      });

      it('sets issuesListLoading to `false`', () => {
        const issuableList = findIssuableList();
        expect(issuableList.props('issuablesLoading')).toBe(false);
      });
    });

    describe('when request fails', () => {
      it.each`
        APIErrorMessage | expectedRenderedErrorMessage
        ${'API error'}  | ${'API error'}
        ${undefined}    | ${'An error occurred while loading issues'}
      `(
        'calls `createFlash` with "$expectedRenderedErrorMessage" when API responds with "$APIErrorMessage"',
        async ({ APIErrorMessage, expectedRenderedErrorMessage }) => {
          jest.spyOn(axios, 'get');
          mock
            .onGet(mockProvide.issuesFetchPath)
            .replyOnce(httpStatus.INTERNAL_SERVER_ERROR, { errors: [APIErrorMessage] });

          createComponent();

          await waitForPromises();

          expect(createFlash).toHaveBeenCalledWith({
            message: expectedRenderedErrorMessage,
            captureError: true,
            error: expect.any(Object),
          });
        },
      );
    });
  });

  it('renders issuable-list component with correct props', async () => {
    createComponent({ initialFilterParams: { search: 'foo' } });

    await waitForPromises();

    const issuableList = findIssuableList();
    expect(issuableList.exists()).toBe(true);
    expect(issuableList.props()).toMatchSnapshot();
  });

  describe('issuable-list events', () => {
    beforeEach(async () => {
      jest.spyOn(axios, 'get');
      createComponent();
      await waitForPromises();
    });

    it('"click-tab" event executes GET request correctly', async () => {
      const issuableList = findIssuableList();

      issuableList.vm.$emit('click-tab', 'closed');
      await waitForPromises();

      expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
        params: {
          labels: undefined,
          page: 1,
          per_page: 2,
          search: undefined,
          sort: 'created_desc',
          state: 'closed',
          with_labels_details: true,
        },
      });
      expect(issuableList.props('currentTab')).toBe('closed');
    });

    it('"page-change" event executes GET request correctly', async () => {
      const mockPage = 2;
      const issuableList = findIssuableList();

      issuableList.vm.$emit('page-change', mockPage);
      await waitForPromises();

      expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
        params: {
          labels: undefined,
          page: mockPage,
          per_page: 2,
          search: undefined,
          sort: 'created_desc',
          state: 'opened',
          with_labels_details: true,
        },
      });
      expect(issuableList.props()).toMatchObject({
        currentPage: mockPage,
        previousPage: mockPage - 1,
        nextPage: mockPage + 1,
      });
    });

    it('"sort" event executes GET request correctly', async () => {
      const mockSortBy = 'updated_asc';
      const issuableList = findIssuableList();

      issuableList.vm.$emit('sort', mockSortBy);
      await waitForPromises();

      expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
        params: {
          labels: undefined,
          page: 1,
          per_page: 2,
          search: undefined,
          sort: 'created_desc',
          state: 'opened',
          with_labels_details: true,
        },
      });
      expect(issuableList.props('initialSortBy')).toBe(mockSortBy);
    });

    it('filter event sets `filterParams` value and calls fetchIssues', async () => {
      const mockFilterTerm = 'foo';
      const issuableList = findIssuableList();

      issuableList.vm.$emit('filter', [
        {
          type: 'filtered-search-term',
          value: {
            data: mockFilterTerm,
          },
        },
      ]);
      await waitForPromises();

      expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
        params: {
          labels: undefined,
          page: 1,
          per_page: 2,
          search: mockFilterTerm,
          sort: 'created_desc',
          state: 'opened',
          with_labels_details: true,
        },
      });
    });
  });

  describe('pagination', () => {
    it.each`
      scenario                 | issuesListLoadFailed | issues            | shouldShowPaginationControls
      ${'fails'}               | ${true}              | ${[]}             | ${false}
      ${'returns no issues'}   | ${false}             | ${[]}             | ${false}
      ${`returns some issues`} | ${false}             | ${mockJiraIssues} | ${true}
    `(
      'sets `showPaginationControls` prop to $shouldShowPaginationControls when request $scenario',
      async ({ issuesListLoadFailed, issues, shouldShowPaginationControls }) => {
        jest.spyOn(axios, 'get');
        mock
          .onGet(mockProvide.issuesFetchPath)
          .replyOnce(
            issuesListLoadFailed ? httpStatus.INTERNAL_SERVER_ERROR : httpStatus.OK,
            issues,
            {
              'x-page': 1,
              'x-total': 3,
            },
          );

        createComponent();

        await waitForPromises();

        expect(findIssuableList().props('showPaginationControls')).toBe(
          shouldShowPaginationControls,
        );
      },
    );
  });
});
