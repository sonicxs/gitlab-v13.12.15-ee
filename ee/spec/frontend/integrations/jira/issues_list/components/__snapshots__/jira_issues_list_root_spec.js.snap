// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraIssuesListRoot renders issuable-list component with correct props 1`] = `
Object {
  "currentPage": 1,
  "currentTab": "opened",
  "defaultPageSize": 2,
  "enableLabelPermalinks": true,
  "initialFilterValue": Array [
    Object {
      "type": "filtered-search-term",
      "value": Object {
        "data": "foo",
      },
    },
  ],
  "initialSortBy": "created_desc",
  "isManualOrdering": false,
  "issuableSymbol": "#",
  "issuables": Array [],
  "issuablesLoading": false,
  "labelFilterParam": "labels",
  "namespace": "gitlab-org/gitlab-test",
  "nextPage": 2,
  "previousPage": 0,
  "recentSearchesStorageKey": "jira_issues",
  "searchInputPlaceholder": "Search Jira issues",
  "searchTokens": Array [],
  "showBulkEditSidebar": false,
  "showPaginationControls": false,
  "sortOptions": Array [
    Object {
      "id": 1,
      "sortDirection": Object {
        "ascending": "created_asc",
        "descending": "created_desc",
      },
      "title": "Created date",
    },
    Object {
      "id": 2,
      "sortDirection": Object {
        "ascending": "updated_asc",
        "descending": "updated_desc",
      },
      "title": "Last updated",
    },
  ],
  "tabCounts": null,
  "tabs": Array [
    Object {
      "id": "state-opened",
      "name": "opened",
      "title": "Open",
      "titleTooltip": "Filter by issues that are currently opened.",
    },
    Object {
      "id": "state-closed",
      "name": "closed",
      "title": "Closed",
      "titleTooltip": "Filter by issues that are currently closed.",
    },
    Object {
      "id": "state-all",
      "name": "all",
      "title": "All",
      "titleTooltip": "Show all issues.",
    },
  ],
  "totalItems": 0,
  "urlParams": Object {
    "labels[]": undefined,
    "page": 1,
    "search": "foo",
    "sort": "created_desc",
    "state": "opened",
  },
}
`;
